# Introduction

The **RiboViz** paper is published: [riboviz: analysis and visualization of ribosome profiling datasets, Carja et al., BMC Bioinformatics 2017](https://doi.org/10.1186/s12859-017-1873-8).

The **RiboViz** web application is available at [riboviz.org](https://riboviz.org). Through this web framework, the user can interactively explore publicly available ribosome profiling datasets using JavaScript/[D3](https://d3js.org/) (Bostock et al., 2011), [JQuery](http://jquery.com) and [Boostrap](http://getbootstrap.com) for metagenomic analyses and [R](https://www.r-project.org/)/[Shiny](https://shiny.rstudio.com/) for the gene-specific analyses. The visualization framework of **RiboViz** allows the user to select from available **RiboSeq** datasets and visualize different aspects of the data.  

The framework allows visualization of the expected three-nucleotide periodicity along the ORFs, accumulation of footprinting reads at the start and stop codons, the distribution of ribosomal-footprint lengths,  the position-specific nucleotide frequencies of all mapped reads, as well as the codon-specific densities of ribosomes. The display also shows the correlation between gene-specific estimates of ribosome densities and various sequence-based features. The user can interactively explore the data and download parsed datasets used to generate each figure. 

The plots also allow the user to compare data with aggregates of other data sets. Ribosome profiling has been shown to exhibit experimental abnormalities that need to be tested for when analyzing a new data set. For example, translocation elongation inhibitors (such as cycloheximide) can alter the local distribution of ribosomes on the mRNA. An advantage of using cycloheximide (CHX) as a  pre-treatment is that it prevents the runoff of ribosomes that can otherwise occur during harvesting. However, this treatment can also have some undesirable effects and can produce aberrant snapshots of where the ribosomes are stalling. This happens especially near the translation start and stop codons (ribosome accumulation at start codons and depletion at stop codons) and it can lead to spurious results. In addition, because CHX binding to the 80S ribosome is both non-instantaneous and reversible, the kinetics of CHX binding and dissociation presumably allow newly initiated ribosomes to translocate beyond the start codon. Another possible effect of CHX treatment is that ribosomes might preferentially arrest at specific codons that do not necessarily correspond to codons that are more abundantly occupied by ribosomes in untreated cells. 

Currently, the literature consists of a mix of studies, some which use cycloheximide and some which use flash freezing. In general, each experimental step can potentially cause spurious results and distortions in the data output. Three of the **RiboViz** visualizations will help identify potential biases in the data by providing the user the ability to compare and contrast different datasets obtained by different experimental conditions.

In addition to the metagenomic analyses, an R/Shiny integration allows the user to select a gene of interest and compare ribosomal densities along its ORF with up to three other data sets. 

## References

(Bostock et al., 2011) Bostock, Michael and Ogievetsky, Vadim and Heer, Jeffrey, 'D3 Data-Driven Documents', IEEE Transactions on Visualization and Computer Graphics, 17(12), pp2301-2309, December 2011, doi: [10.1109/TVCG.2011.185](http://dx.doi.org/10.1109/TVCG.2011.185).
